tests/test_training/test_opt.py

```diff
@@ -457,3 +457,14 @@ def cost_fn():
 
     assert np.allclose(bsgate.theta.value, 0.1, atol=0.01)
     assert np.allclose(bsgate.phi.value, 0.2, atol=0.01)
+
+
+def test_squeezing_grad_from_fock():
```

Comment:
Should we test the other gates that have custom gradients too?

Reply:
This worked.

`from mrmustard.training import Optimizer, Parametrized
from mrmustard.lab import *

sq = SqueezedVacuum(r=1.0,r_trainable=True)

def cost_fn():
    return -((sq &Fock(1)) >> BSgate(0.5) << (Vacuum(1)&Fock(1)))

opt = Optimizer(euclidean_lr=0.05)
opt.minimize(cost_fn, by_optimizing=[sq], max_steps=100)`

Reply:
This did not.

`from mrmustard.training import Optimizer, Parametrized
from mrmustard.lab import *

disp = Dgate(x=1.0,y=1.0, x_trainable=True,y_trainable=True)

def cost_fn():
    return -(Fock(2)>> disp << Vacuum(1))

opt = Optimizer(euclidean_lr=0.05)
opt.minimize(cost_fn, by_optimizing=[disp], max_steps=100)`

Reply:
This worked.

`from mrmustard.training import Optimizer, Parametrized
from mrmustard.lab import *

bs = BSgate(theta=0.5,theta_trainable=True)

def cost_fn():
    return -((Fock(2,cutoffs=[3]) &Fock(1,cutoffs=[3])) >> bs << (Fock(2,cutoffs=[3]) &Fock(1,cutoffs=[3])))

opt = Optimizer(euclidean_lr=0.05)
opt.minimize(cost_fn, by_optimizing=[bs], max_steps=100)`

Reply:
caught it!